### ISSUE TITLE
Proposal to add --prefer-ac3 option to transcode-video

### ISSUE THREAD

**Issue by donmelton** [donmelton/video_transcoding#112]
# Proposal to add `--prefer-ac3` option to `transcode-video`

No denies that Dolby Digital AC-3 audio is a legacy format still with us from the early 1990s and inefficient compared to the storage requirements of AAC. However, it's possible that even your toaster can decode AC-3. That's just how ubiquitous the format is.

While HandBrake includes an excellent AC-3 encoder, the same cannot be said of AAC unless you're fortunate enough to be on macOS where Core Audio is available.

The `transcode-video` tool currently defaults to encoding or copying the main audio track in AC-3 format as long as the original track layout is greater than 2.0 channels.

Unfortunately, there's no easy way using `transcode-video` to prefer AC-3 over AAC format when the original track channel layout is stereo or mono.

Here's how I would describe a new option, `--prefer-ac3`, designed to change that.

```
    --prefer-ac3    prefer AC-3 over AAC format when encoding or copying audio
                      even when original track channel layout is stereo or mono
                      (sets audio output "width" for all tracks to `surround`)
```

Would anyone else (besides me) use this?

Let me know what you think. Soon. Because I'm liable to add the option sometime this week.

Thanks.

**Comment by bmhayward** [comment]
Been wanting this for a while. Ship it!

Thanks,
Brent

**Comment by donmelton** [comment]
@bmhayward Wow! I had no idea. :)

**Comment by donmelton** [comment]
BTW, here's the patch to implement the feature:

```
diff --git a/bin/transcode-video b/bin/transcode-video
index d83f509..ed37ea1 100755
--- a/bin/transcode-video
+++ b/bin/transcode-video
@@ -124,6 +124,9 @@ Audio options:
                       with `surround` to allow single surround or stereo track
                       with `stereo` to allow only single stereo track
                       (can be used multiple times)
+    --prefer-ac3    prefer AC-3 over AAC format when encoding or copying audio
+                      even when original track channel layout is stereo or mono
+                      (sets audio output "width" for all tracks to `surround`)
     --ac3-bitrate 384|448|640
                     set AC-3 audio bitrate (default: 640)
     --pass-ac3-bitrate 384|448|640
@@ -239,6 +242,7 @@ HERE
       @audio_name                 = {}
       @audio_language             = []
       @audio_width                = {:main => :double, :other => :stereo}
+      @prefer_ac3                 = false
       @ac3_bitrate                = 640
       @pass_ac3_bitrate           = 640
       @copy_audio                 = []
@@ -533,6 +537,12 @@ HERE
         end
       end
 
+      opts.on '--prefer-ac3' do
+        @prefer_ac3 = true
+        @audio_width[:main] = :surround
+        @audio_width[:other] = :surround
+      end
+
       opts.on '--ac3-bitrate ARG', Integer do |arg|
         @ac3_bitrate = case arg
         when 384, 448, 640
@@ -1090,7 +1100,7 @@ HERE
             add_stereo.call info, copy
           end
         when :surround
-          if info[:channels] > 2.0
+          if @prefer_ac3 or (info[:channels] > 2.0)
             add_surround.call info, copy
           else
             add_stereo.call info, copy
```

As you can see, the implementation is trivial. :)

**Comment by RodBrown1988** [comment]
Currently the current defaults work for me. Considering AC 3 is legacy and I'm running on an iMac, I don't see any point myself in switching from the default.

That said, you make good points about other platforms and about compatibility. I think this is a reasonable option. Thumbs up!

**Comment by donmelton** [comment]
@RodBrown1988 Yeah, I didn't think folks would be rushing to use this, especially on macOS. :) Thanks for the feedback!

**Comment by ericcardenas** [comment]
Yes please. 
AC3 is still standard even if it's legacy stuff.

**Comment by samhutchins** [comment]
Doubt I'd use it, I compile my own handbrake with fdk on Windows

That said, I see no harm in adding the option :-P

**Comment by donmelton** [comment]
@samhutchins Yeah, compiling in the Fraunhofer FDK AAC library on Windows and Linux is currently the only way to get parity with macOS. Smart move. And thanks for the feedback!

**Comment by timsutton** [comment]
I think this option would have helped me in a recent batch of encodes I did, where it took me a while to figure out the needed set of --copy and --add audio options to preserve the original AC3 track instead of re-encoding it, which I wanted to avoid.

Would definitely use this!

**Comment by donmelton** [comment]
@timsutton Cool. Sounds like you were running into the same problem I was. :)

**Comment by martinpickett** [comment]
I would also use this feature. I'd probably not use it often, but it would be useful.

**Comment by donmelton** [comment]
@martinpickett Thanks for the feedback! Likely it will get released tonight or tomorrow.

**Comment by timsutton** [comment]
In general, I prefer to always keep AC3 because of exactly what you stated, which is that it's so universally supported for decoding and there isn't much bitrate savings to be had in re-compressing.

**Comment by donmelton** [comment]
@timsutton Yes, and considering that commentary tracks (which I always add) are usually 192 Kbps AC-3, re-encoding them as 160 Kbps AAC seems pretty silly. :)

**Comment by cameronks** [comment]
Don, I would definitely use this. I used to be Mac only, but I built a server on Windows this past spring and started using `--copy-audio all --add-audio all ` on some of my AC-3 only files for the same reasons you mention.

**Comment by donmelton** [comment]
@bmhayward @RodBrown1988 @ericcardenas @samhutchins @timsutton @martinpickett OK, this has been released (along with an unrelated but very important bug fix) as version 0.14.0. Just `gem update video_transcoding` to get it all.

And thanks again for all your feedback and support!
